core/include/detray/tools/surface_factory.hpp

```diff
@@ -0,0 +1,339 @@
+      public std::enable_shared_from_this<surface_factory<
+          detector_t, mask_shape_t, mask_id, sf_id, volume_link_t>> {
+    public:
+    using scalar_t = typename detector_t::scalar_type;
+    // Set individual volume link for portals, but only the mothervolume index
+    // for other surfaces.
+    using volume_link_collection =
+        std::conditional_t<sf_id == surface_id::e_portal,
+                           std::vector<volume_link_t>,
+                           detray::views::single<volume_link_t>>;
+    // ensures correct returning of shared pointers without pointless
+    // construction and destructing of surface factories
+    using smart_ptr_handler = std::enable_shared_from_this<surface_factory<
+        detector_t, mask_shape_t, mask_id, sf_id, volume_link_t>>;
+
+    /// shorthad for a colleciton of surface data that can be read by a surface
+    /// factory
+    using sf_data_collection = std::vector<std::unique_ptr<
+        surface_data_interface<detector_t, mask_shape_t, volume_link_t>>>;
+
+    /// Empty factory.
+    DETRAY_HOST
+    surface_factory() = default;
+
+    /// @returns the current number of surfaces that will be built by this
+    /// factory
+    DETRAY_HOST
+    auto size() const -> std::size_t {
+
+        // This should not happend (need same number and ordering of data)
+        assert(m_components.size() == m_transforms.size());
+        if constexpr (sf_id == surface_id::e_portal) {
+            assert(m_components.size() == m_volume_link.size());
+        } else {
+            assert(m_volume_link.size() == 1);
+        }
+
+        return m_components.size();
+    }
+
+    /// @returns the mask boundaries currently held by the factory
+    DETRAY_HOST
+    auto components() const -> const std::vector<std::vector<scalar_t>> & {
+        return m_components;
+    }
+
+    /// @returns the transforms currently held by the factory
+    DETRAY_HOST
+    auto transforms() const
+        -> const std::vector<typename detector_t::transform3> & {
+        return m_transforms;
+    }
+
+    /// @returns the volume link(s) currently held by the factory
+    DETRAY_HOST
+    const auto &volume_links() const { return m_volume_link; }
+
+    /// Add all necessary compontents to the factory from bundled surface
+    /// data in @param surface_data .
+    DETRAY_HOST
+    auto add_components(sf_data_collection &&surface_data)
+        -> std::shared_ptr<surface_factory<detector_t, mask_shape_t, mask_id,
+                                           sf_id, volume_link_t>> {
+
+        m_transforms.reserve(surface_data.size());
+        m_components.reserve(surface_data.size());
+
+        if constexpr (sf_id != surface_id::e_portal) {
+            *m_volume_link = std::get<1>(surface_data[0]->get_data());
+        } else {
+            m_volume_link.reserve(surface_data.size());
+        }
+
+        // Get AoS surface data into detector level SoA-like layout
```

Comment:
What is SoA-like here?

Reply:
I meant moving from data per surface to the surfaces begin distributed over multiple contiguous containers in the detector. Maybe that's misleading here...